lib/eth/EthRpc.js

```diff
@@ -177,56 +178,184 @@ class EthRPC {
+    return this._estimateFee({
+      cache: this.blockGasPriceCache,
+      rpcEstimate,
+      percentile: 25,
+    });
+  }
+
+  /**
+   * Estimates the maximum base fee. Defaults to 2 * baseFee + 2.5 gwei.
+   * @param {Object} params - The parameters for estimating the maximum base fee.
+   * @param {number} [params.percentile] - Optional: The maxPriorityFee percentile from last block to use for the estimation.
+   * @param {number} [params.priority] - Optional: The maxPriorityFee to be used with the baseFee.
+   * @returns {number} The estimated maximum base fee.
+   */
+  async estimateMaxFee({ percentile,  priority }) {
+    const lastBlock = await this.web3.eth.getBlock('latest');    
+    const baseFeePerGas = parseInt(lastBlock.baseFeePerGas) ? parseInt(lastBlock.baseFeePerGas) : 0;
+
+    if (priority && parseInt(priority) > 0) {
+      // user defined priority fee
+      return baseFeePerGas + priority;
+    }
+    if (percentile && parseInt(percentile) > 0 && parseInt(percentile) <= 100) {
+      // percentile bases priority fee
+      const maxPriorityFee = await this.estimateMaxPriorityFee({percentile});
+      return baseFeePerGas + maxPriorityFee;
+    }
+    // default max fee formmula. ensures inclusion in next block
+    return 2 * baseFeePerGas + parseInt(this.web3.utils.toWei('2.5', 'gwei'));
+  }
+
+  /**
+   * Estimates the maximum priority fee.
+   * @param {Object} params - The parameters for estimating the maximum priority fee.
+   * @param {number} [params.percentile=25] - The percentile to use for the estimation.
+   * @returns {Promise<number>} The estimated maximum priority fee.
+   */
+  async estimateMaxPriorityFee({ percentile = 25 }) {
+    const rpcEstimate = parseInt(this.web3.utils.toWei('2.5', 'gwei'));
+    const getFees = (_txs) => _txs
+      .filter((tx) => !!tx.maxPriorityFeePerGas)
+      .map((tx) => parseInt(tx.maxPriorityFeePerGas));
+    return this._estimateFee({
+      cache: this.blockMaxPriorityFeeCache,
+      rpcEstimate,
+      percentile,
+      getFees,
+    });
+  }
+
+  /**
+   * Fetches the last n blocks from the blockchain.
+   * @param {Object} params - The parameters for fetching the blocks.
+   * @param {Map} params.cache - The cache to use for storing the blocks.
+   * @param {number} [params.n=10] - The number of blocks to fetch.
+   * @returns {Promise<Array>} The fetched blocks.
+   */
+  async _getRecentBlocks({ cache, n = 10 }) {
     const bestBlock = await this.web3.eth.getBlockNumber();
-    const blocks = await Promise.all([...Array(10).keys()]
-      .map((n) => {
-        let targetBlockNumber = bestBlock - n;
-        if (this.blockGasPriceCache.has(targetBlockNumber)) {
-          return this.blockGasPriceCache.get(targetBlockNumber);
+    const blocks = await Promise.all(
+      [...Array(n).keys()].map((n) => {
+        const targetBlockNumber = bestBlock - n;
+        if (cache && cache.has(targetBlockNumber)) {
```

Comment:
This can cause a bug if different percentiles are queried. The block number may be in the cache but not necessarily  containing the right percentile. Without the block data we can't create the right percentile. Not sure if a fix is as simple as checking for the right percentile here or not. This should be fixed if we make this method only return blocks.